src/hazelcore/extractions/ocaml/OCamlExtraction_Pat.re

```diff
@@ -0,0 +1,114 @@
+// mark with a constructor to tell apart with typ, exp
+type t =
+  | OCamlPat(string);
+
+exception Pat_Hole;
+exception Pat_Keyword(ExpandingKeyword.t); //there's Keyword
+exception Pat_Invalid(string);
+
+// the ocaml naming standard can be found in Chapter 7.3 of ocaml-manual (https://caml.inria.fr/pub/docs/manual-ocaml/names.html)
+let ocaml_name_check = (name: Var.t): bool => {
+  let re = Re.Str.regexp("^[_a-z][_a-zA-Z0-9]*$");
+  Re.Str.string_match(re, name, 0);
+};
+
+let rec extract = (dp: DHPat.t): t =>
+  switch (dp) {
+  | EmptyHole(_, _)
+  | NonEmptyHole(_, _, _, _) => raise(Pat_Hole)
+  | Wild => OCamlPat("_")
+  | Keyword(_, _, keyword) => raise(Pat_Keyword(keyword))
+  | InvalidText(_, _, s) => raise(Pat_Invalid("text \"" ++ s ++ "\""))
+  | Var(s) =>
+    //check variable naming is valid
+    if (ocaml_name_check(s)) {
+      OCamlPat(s);
+    } else {
+      raise(Pat_Invalid("variable naming \"" ++ s ++ "\""));
+    }
+  | IntLit(i) => OCamlPat(string_of_int(i))
+  | FloatLit(f) => OCamlPat(string_of_float(f))
+  | BoolLit(b) => OCamlPat(string_of_bool(b))
+  // As the type design, we encode it as "|`Left l | `Right r"
+  // injection in pattern is used in pattern matching, injL matches the left side of type
+  | Inj(side, t) =>
+    switch (extract(t)) {
+    | OCamlPat(pat) =>
+      switch (side) {
+      | L => OCamlPat("(`Left " ++ pat ++ ")")
+      | R => OCamlPat("(`Right " ++ pat ++ ")")
+      }
+    }
+  | ListNil => OCamlPat("[]")
+  | Cons(dp1, dp2) =>
+    switch (extract(dp1), extract(dp2)) {
+    | (OCamlPat(pat1), OCamlPat(pat2)) =>
+      OCamlPat("(" ++ pat1 ++ "::" ++ pat2 ++ ")")
+    }
+  | Pair(dp1, dp2) =>
+    switch (extract(dp1), extract(dp2)) {
+    | (OCamlPat(pat1), OCamlPat(pat2)) =>
+      OCamlPat("(" ++ pat1 ++ ", " ++ pat2 ++ ")")
+    }
+  | Triv => OCamlPat("()")
+  | Ap(dp1, dp2) =>
```

Comment:
What cases cause this?  Are there constructors that are applied?  (`Inj` already handles `Left` and `Right'.)

Reply:
Sorry I can't understand that, is that the case of `Ap`?

Reply:
Yes, what causes an `Ap` pattern?

Reply:
it can exist syntactically because we need a space operator for editing reasons (when you delete another operator you need to leave something behind), but it will always be in a hole.

Reply:
Holes will cause an extraction failure, so I think it's better to remove it because it will never be used.

Reply:
Oh, I want to preserve it now. The reason is the `extract` function should return a string or raise an exception, but it seems really strange to raise some exception because it's definitely not an exception. Meanwhile, since it can exists syntactically, I don't think it will cause problem.

Reply:
I'm starting to think it should be an exception because it is a form that extraction does not support, just like extraction doesn't support holes.  Since this case should never occur, maybe a `failwith __LOC__` or some other exception that represents "impossible" cases in the code.  (Note that OCaml `(p1 p2)`, which is what it currently extracts to, does not have the same meaning as Hazel `(p1 p2)`.)

Reply:
I'm combining the `extract` and `update_pattern`, so here I can throw an exception of "invalid update". Meanwhile, can you explain what's the difference between OCaml `p1 p2` and Hazle `p1 p2`? I thought they're both apply.

Reply:
Done, now the `OCamlExtraction_Pat.extract` will directly throw an exception on `Ap` case.

Reply:
Yeah it's fine to throw an exception for this. When we have algebraic datatypes we will need this form.